rust/core-lib/src/layers.rs

```diff
@@ -0,0 +1,171 @@
+    }
+
+    /// Updates spans on all layers. Useful for clearing all, or OT.
+    pub fn update_all(&mut self, iv: Interval, spans: Spans<u32>) {
+        for layer in self.layers.values_mut() {
+            layer.update_spans(iv, spans.clone());
+        }
+    }
+
+    /// Updates the scope spans for a given layer.
+    pub fn update_layer(&mut self, layer: &str, iv: Interval, spans: Spans<u32>) {
+        self.create_if_missing(layer);
+        self.layers.get_mut(layer).unwrap().update_spans(iv, spans);
+    }
+
+    /// Removes a given layer. This will remove all styles derived from
+    /// that layer's scopes.
+    pub fn remove_layer(&mut self, layer: &str) -> Option<ScopeLayer> {
+        self.layers.remove(layer)
+    }
+
+    /// For a given Interval, generates styles from scopes, resolving conflicts.
+    pub fn resolve_styles(&self, iv: Interval) -> Spans<Style> {
+        if self.layers.is_empty() {
+            return SpansBuilder::new(iv.size()).build()
+        }
+        //TODO: implement layer merging.
+        let mut sb = SpansBuilder::new(iv.size());
+        let layer = self.layers.values().next().unwrap();
+        for (iv, val) in layer.spans.subseq(iv).iter() {
+            let style = layer.style_lookup[*val as usize];
+            sb.add_span(iv, style);
+        }
+        sb.build()
+    }
+
+    pub fn debug_print_spans(&self, iv: Interval) {
+        for (id, layer) in self.layers.iter() {
+            let spans = layer.spans.subseq(iv);
+            print_err!("Spans for layer {}:", id);
+            for (iv, val) in spans.iter() {
+                print_err!("{}: {:?}", iv, layer.name_lookup[*val as usize])
+            }
+        }
+    }
+
+    fn create_if_missing(&mut self, layer_id: &str) {
+        if !self.layers.contains_key(layer_id) {
+            self.layers.insert(layer_id.to_owned(), ScopeLayer::default());
+        }
+    }
+}
+
+impl Default for Scopes {
+    fn default() -> Self {
+        Scopes { layers: BTreeMap::new() }
+    }
+}
+
+impl Default for ScopeLayer {
+    fn default() -> Self {
+        ScopeLayer {
+            stack_lookup: Vec::new(),
+            style_lookup: Vec::new(),
+            name_lookup: Vec::new(),
+            spans: Spans::default(),
+        }
+    }
+}
+
+impl ScopeLayer {
+    fn add_scopes<W: Write>(&mut self, scopes: Vec<Vec<String>>,
+                                doc_ctx: &DocumentCtx<W>) {
+        let mut stacks = Vec::with_capacity(scopes.len());
+        for stack in scopes {
+            let scopes = stack.iter().map(|s| Scope::new(&s))
+                .filter(|result| match *result {
+                   Err(ref err) => {
+                       print_err!("failed to resolve scope {}\nErr: {:?}",
+                                  &stack.join(" "),
+                                  err);
+                       false
+                       }
+                   _ => true
+                })
+            .map(|s| s.unwrap())
```

Comment:
this indentation seems weird. I think this and the other iterator call below should be indented just like `filter` above.